### PR TITLE
Add new config param to allow GroupExec perm for the SecretBackend command

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

To allow the execution of the secret-backend command packaged inside
the docker image. the secret-backend script need to allow the execution
of a group uid.
The new option "secret_backend_command_allow_group_exec_perm" allow to
change the behaviour of the `checkRights()` function and allow GroupExecPerm.

We update only the dockerfile configuration for the cluster-agent. because
the "agent" process needs to run as root, so the described setup is not possible
with the agent image:
* change the readsecret.sh permission to 510
* set DD_SECRET_BACKEND_COMMAND_ALLOW_GROUP_EXEC_PERM env var to true
  by default in the cluster-agent entry-point

### Motivation

When the agent runs in a container environment, the process can be
executed with a "random" uid (for security reason)

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Start the `Cluster-Agent` and use the "secret-backend" script packaged in the image `readSecret.py`. The `Cluster-Agent` should not return a file permission check error.

### CODE REVIEW

File: Dockerfiles/cluster-agent/amd64/Dockerfile
Comment (L3n41c) [inline Dockerfiles/cluster-agent/amd64/Dockerfile:19]: The “execute only” bit on the shell script is useless because, on the contrary to an elf binary, in order to be executable, a script needs to be both executable and readable.

Demo:
```
vagrant@agent-dev-lenaic:/tmp$ ls -l hello hello.sh
-r-x--x--x 1 root root 16696 Aug 31 07:52 hello
-r-x--x--x 1 root root    42 Aug 31 07:51 hello.sh

# ELF binary
vagrant@agent-dev-lenaic:/tmp$ ./hello
Hello

# bash script
vagrant@agent-dev-lenaic:/tmp$ ./hello.sh
/bin/bash: ./hello.sh: Permission denied
```

This is because the executable bit is needed to start the script and the readable bit is needed for the shell interpreter to be allowed to read the script.

```suggestion
    && chmod 550 readsecret.sh \
```

**Review by L3n41c — CHANGES_REQUESTED** [review]

File: pkg/secrets/check_rights_nix_test.go
Comment (L3n41c) [inline pkg/secrets/check_rights_nix_test.go:43]:
```suggestion
	// owner have write permission
```

File: pkg/secrets/check_rights_nix_test.go
Comment (L3n41c) [inline pkg/secrets/check_rights_nix_test.go:59]:
```suggestion
	// group should have only read and exec rights
	require.Nil(t, os.Chmod(tmpfile.Name(), 0750))
	require.Nil(t, checkRights(tmpfile.Name(), groupOptions))
```

File: pkg/secrets/check_rights_nix_test.go
Comment (L3n41c) [inline pkg/secrets/check_rights_nix_test.go:63]:
```suggestion
	// group should not have write right
	require.Nil(t, os.Chmod(tmpfile.Name(), 0770))
	require.NotNil(t, checkRights(tmpfile.Name(), groupOptions))
```

File: pkg/secrets/no_secrets.go
Comment (L3n41c) [inline pkg/secrets/no_secrets.go:20]: Group Exec **Permission** ?

```suggestion
func Init(command string, arguments []string, timeout int, maxSize int, groupExecPerm bool) {}
```

File: pkg/secrets/secrets.go
Comment (L3n41c) [inline pkg/secrets/secrets.go:42]:
```suggestion
func Init(command string, arguments []string, timeout int, maxSize int, groupExecPerm bool) {
```

File: pkg/secrets/secrets.go
Comment (L3n41c) [inline pkg/secrets/secrets.go:47]:
```suggestion
	secretBackendCommandAllowGroupExec = groupExecPerm
```

**Review by L3n41c — CHANGES_REQUESTED** [review]
The biggest issue is the lack of read permission for a script.

Otherwise, I left some cosmetics suggestions.

File: pkg/secrets/check_rights_nix.go
Comment (L3n41c) [inline pkg/secrets/check_rights_nix.go:32]: Maybe the message is a little bit misleading.

```suggestion
		return fmt.Errorf("invalid executable '%s' permissions, current file permissions are %#o but %#o are unwanted", path, stat.Mode, unwantedPermissions)
```

**Review by L3n41c — APPROVED** [review]
There is only one error message which is misleading.

File: pkg/secrets/check_rights_nix.go
Comment (hush-hush) [inline pkg/secrets/check_rights_nix.go:25]: Could we update the comment to reflect what we're blacklisting ? Something similar to `Allow group to have read and exec rights, blacklisting write`

File: pkg/secrets/check_rights_nix.go
Comment (hush-hush) [inline pkg/secrets/check_rights_nix.go:31]: Is the `uint` cast needed ?

File: pkg/secrets/check_rights.go
Comment (hush-hush) [inline pkg/secrets/check_rights.go:11]: I don't see any other options for the check rights (and this is the first change we made in a long time on this). I feel this is not needed and a simple `bool` in the `checkRights` signature would be enough and keep the code simpler.

File: pkg/secrets/check_rights_windows.go
Comment (hush-hush) [inline pkg/secrets/check_rights_windows.go:27]: Let's make it clear options are not used for the windows version:
```suggestion
func checkRights(filename string, _ checkRightOptions) error {
```

File: releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (hush-hush) [inline releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:4]: Could we develop this to quote the new parameter name and explain in which case it should be enabled ?

File: pkg/secrets/secrets.go
Comment (hush-hush) [inline pkg/secrets/secrets.go:47]: Could we log something when this is set to true ? Do we want to return on error when we don't run in a containerized env ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
A few comments. But overall if security is OK with this I think we're almost good to merge.

**Comment by hush-hush** [comment]
Once this is merged let's not forget to update the official doc to explain when, why and how to enable this option : https://docs.datadoghq.com/agent/guide/secrets-management/?tab=linux

File: pkg/secrets/check_rights_nix.go
Comment (hush-hush) [inline pkg/secrets/check_rights_nix.go:32]: Also could we improve the error message ? For example ` current file permissions are 0100640 but 027 are unwanted` is pretty obscure and not very useful for customer.

We could have 2 error messages and use the right one based on `AllowGroupExec`

File: pkg/secrets/check_rights_nix.go
Comment (hush-hush) [inline pkg/secrets/check_rights_nix.go:23]: Why not:
```suggestion
	unwantedPermissions := syscall.S_IRWXO
```

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/secrets/check_rights_nix.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go:31]: It was to be able to compare with `var unwantedPermissions uint = syscall.S_IRWXO`

File: pkg/secrets/check_rights.go
Comment (clamoriniere) [reply pkg/secrets/check_rights.go:11]: Yes, I was divided between simplicity and be future proof.

File: pkg/secrets/check_rights_nix.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go:23]: when I did that, the linter was failing in some ci build. because the syscall.S_IRWXO was define as different type.

File: pkg/secrets/secrets.go
Comment (clamoriniere) [reply pkg/secrets/secrets.go:47]: I think the log is not properly configure when we call the `Init()` function the `secret` package.
If we do at exec time, it will be quite verbose if they are decoding lot of secrets.

File: pkg/secrets/check_rights_nix.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go:31]: any way I will remove the usage of `unwantedPermissions ` by having 2 different error messages.

File: pkg/secrets/check_rights_nix.go
Comment (olivielpeau) [inline pkg/secrets/check_rights_nix.go:26]:
```suggestion
			return fmt.Errorf("invalid executable '%s', 'others' have rights on it or 'group' has write permissions on it", path)
```

File: pkg/secrets/check_rights_nix.go
Comment (olivielpeau) [inline pkg/secrets/check_rights_nix.go:31]:
```suggestion
			return fmt.Errorf("invalid executable '%s', 'group' or 'others' have rights on it", path)
```

File: pkg/secrets/secrets.go
Comment (olivielpeau) [reply pkg/secrets/secrets.go:47]: it should be fine to `log` before the log package is initialized, when that happens the `log` package buffers the logs and actually flushes them once it's been initialized.

Logging a warning in non-containerized envs would make sense to me

File: releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (olivielpeau) [inline releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:8]: This note should probably be in a cluster-agent release note, not an agent one

**Review by olivielpeau — APPROVED** [review]
Left a few nits inline, once addressed LGTM 👍 

Let's have a security review as well

File: pkg/secrets/check_rights_nix.go
Comment (trishankatdatadog) [inline pkg/secrets/check_rights_nix.go:25]: I'm not sure this is what you meant. Did you mean this instead?

`stat.Mode&(syscall.S_IWGRP|syscall.S_IRWXO) != 0`

Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go]: good catch, 
I messed up the 2 bit masks in the `if else`. Should be good now.

File: pkg/secrets/check_rights_nix.go
Comment (trishankatdatadog) [inline pkg/secrets/check_rights_nix.go:30]: Hmm, not sure this is what you meant either. Did you mean `syscall.S_IRWXG|syscall.S_IRWXO`?

Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go]: fixed

File: pkg/secrets/check_rights_nix.go
Comment (trishankatdatadog) [reply pkg/secrets/check_rights_nix.go]: Now this bug is back after your force-push...

File: pkg/secrets/check_rights_nix_test.go
Comment (trishankatdatadog) [inline pkg/secrets/check_rights_nix_test.go:39]: Can we execute w/o being able to read?

File: pkg/secrets/check_rights_nix.go
Comment (trishankatdatadog) [inline pkg/secrets/check_rights_nix.go:23]: How would the user ID check below work with `allowGroupExec`?

File: pkg/secrets/check_rights_nix_test.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix_test.go:39]: no, we need both.

File: pkg/secrets/check_rights_windows.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_windows.go:27]: This option will be mainly used in `cluster-agent` and we release the cluster-agent for linux.
Since the agent still need to run as root, this option is not very useful.

File: pkg/secrets/check_rights_nix_test.go
Comment (trishankatdatadog) [reply pkg/secrets/check_rights_nix_test.go:39]: So perhaps it's time to update the `checkRights()` function and the tests, then?

File: pkg/secrets/check_rights_windows.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_windows.go:27]: I have added a comment in the file.

File: releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (hush-hush) [inline releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:1]: I guess we should remove this file this the other release note include it.

File: releasenotes-dca/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (hush-hush) [inline releasenotes-dca/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:8]: Could we explain what was fixed ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
I think it's almost good to go for me.

File: releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (clamoriniere) [reply releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:1]: One is for the "agent", the other one is for the "DCA". 

Since the feature was design for the DCA, maybe we can remove it from the Agent release notes. Is that ok? event if the option is visible in the agent config?

File: releasenotes-dca/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (clamoriniere) [reply releasenotes-dca/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml]: done

File: pkg/secrets/check_rights_nix.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go:23]: we don't remove owner permission. so the [check](https://github.com/DataDog/datadog-agent/pull/6298/files#diff-84a4058c436ae98c7c84d1b1bddee17cR36) should work

File: pkg/secrets/check_rights_nix_test.go
Comment (clamoriniere) [inline pkg/secrets/check_rights_nix_test.go:40]:
```suggestion
	// we should at least be able to read and execute it
	require.Nil(t, os.Chmod(tmpfile.Name(), 0500))
```

File: pkg/secrets/check_rights_nix_test.go
Comment (trishankatdatadog) [reply pkg/secrets/check_rights_nix_test.go]: I see the code tests `0500` now, but there is no fix or test that `0100` alone is not enough. Up to you guys whether you want to fix this: it is not a security concern, just a failure to read _and_ execute.

File: pkg/secrets/check_rights_nix.go
Comment (trishankatdatadog) [reply pkg/secrets/check_rights_nix.go:23]: I understand, but why would this allow a group to execute something even if it is group-executable but _not_ owned by the current user in that group? Again, not a security issue, but a usability one.

Comment (clamoriniere) [reply pkg/secrets/check_rights_nix.go:23]: it is to follow this openshift guideline about "Support arbitrary user ids": https://docs.openshift.com/container-platform/4.1/openshift_images/create-images.html#images-create-guide-openshift_create-images

File: pkg/secrets/check_rights_nix_test.go
Comment (clamoriniere) [reply pkg/secrets/check_rights_nix_test.go]: Did the change too fast:
* If the "command" is a binary `0100` is enough
* but if it is a script `bash`, `python`... `0500` is needed. 
So I will put back: 0100 in the test

File: releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml
Comment (hush-hush) [reply releasenotes/notes/Add-Group-exec-perm-for-the-SecretBackend-command-a7fbe1ece1fad50b.yaml:1]: I missed the difference. Let's keep both then

**Review by hush-hush — APPROVED** [review]

File: pkg/secrets/check_rights_nix.go
Comment (hush-hush) [inline pkg/secrets/check_rights_nix.go:74]:
```suggestion
			return fmt.Errorf("invalid executable: '%s' isn't owned by this user or one of his group: username '%s', UID %s GUI %s. We can't execute it", path, usr.Username, usr.Uid, usr.Gid)
```

**Review by hush-hush — APPROVED** [review]
One nit but looks good otherwise

File: pkg/secrets/check_rights_nix_test.go
Comment (olivielpeau) [inline pkg/secrets/check_rights_nix_test.go:59]:
```suggestion
	// even if allowGroupExec=true, group may have no permission
```

**Review by olivielpeau — APPROVED** [review]
Nits, and one comment on the logged warning, but the logic LGTM! 👍  (and thanks for the additional unit tests!)

File: pkg/secrets/check_rights_nix_test.go
Comment (trishankatdatadog) [inline pkg/secrets/check_rights_nix_test.go:71]: Consider adding a separate `0702` test here (others can write)

**Review by trishankatdatadog — APPROVED** [review]
Other than warning that a script is executable but not readable (very minor comment), and adding tests for checking UID, I see no security issue. Thanks for all your work!

**Pull request merged**
